src/Compiler/CMakeLists.txt

```diff
@@ -60,7 +60,7 @@ if (ONNX_MLIR_ENABLE_JNI)
   add_dependencies(CompilerUtils jniruntime)
 endif()
 
-add_onnx_mlir_library(OnnxMlirCompiler SHARED
+add_onnx_mlir_library(OnnxMlirCompiler
```

Comment:
Why is this no longer a shared library? This was intentional when the library was created, so that it can be used by other (external) components as a shared library even when the rest of the libraries are built statically

Reply:
A library doesn't have to be shared in order to be used by other components. Either a .a or a .so can be used by other components. Being shared simply means the symbols get resolved at runtime by the loader, while being static means the symbols get resolved at compile time by the linker.

LLVM builds all its libraries shared or static, depending on whether BUILD_SHARED_LIBS is ON or OFF. That's what ONNX-MLIR does too, i.e., all libraries should honor the BUILD_SHARED_LIBS option. The exception is libcruntime.a. It's always built as static because its symbols need to be embedded into the model.so.

Reply:
Shared libraries and static libraries are used differently, especially DLLs on Windows. This library was added explicitly to be a shared library on all platforms, including Windows.

LLVM builds *most* of its libraries as either shared or static, but there are several similar libraries that are always either static or shared because they are special. This was the PR: https://github.com/onnx/onnx-mlir/pull/894

Reply:
Yes I know it was added by PR#984. I just don't see any particular reason why it has to be shared if all you want is  other components can use it. "shared" is a somewhat misleading word. A more appropriate word is "dynamic". Static libraries are also shared. They are just shared at compile time rather than at load time.

Those special LLVM libraries that are always built static or shared must have reasons beyond simply being callable from other components. Like our libcruntime.a I mentioned, it must be always built static because its symbols need to be embedded into the model.so at compile time to make model.so self-contained.

The original words from @MikeHolman are "With a shared library we can simply give them the dll and not require them to add onnx-mlir and all its dependencies (in particular LLVM/MLIR) to their build." 

Perhaps a DLL is required on Windows but on other platforms a static lib would work fine as far as I can tell. If that's the case, we can special case for Windows. But on other platforms, OnnxMlirCompiler should honor BUILD_SHARED_LIBS option.